File: launcher.js
const { spawn, exec } = require('child_process');
const path = require('path');
const fs = require('fs');

// Cấu hình tham số môi trường
const PORT = process.env.PORT || 3000;
process.env.PORT = PORT;
process.env.NODE_ENV = 'production';
process.env.HOSTNAME = '127.0.0.1'; // Giới hạn chỉ chạy loopback tránh cảnh báo tường lửa Windows

const rootDir = __dirname;
const nodeExe = path.join(rootDir, 'bin', 'node.exe');
const serverJs = path.join(rootDir, 'server.js');

if (!fs.existsSync(serverJs)) {
  console.error("Lỗi: Không tìm thấy file server.js trong thư mục local!");
  console.error("Vui lòng đảm bảo cấu trúc thư mục chính xác.");
  readlineHelper();
  return;
}

// Kiểm tra node.exe portable
const nodePath = fs.existsSync(nodeExe) ? nodeExe : 'node';

console.log("=========================================================");
console.log("          K H Ở I   Đ Ộ N G   T E C H P R O J E C T      ");
console.log("=========================================================");
console.log(`  Ứng dụng đang được khởi chạy tại: http://localhost:${PORT}`);
console.log("  Vui lòng GIỮ cửa sổ này luôn mở khi sử dụng phần mềm.");
console.log("  Để dừng ứng dụng: Nhấn phím Ctrl + C trong cửa sổ này.");
console.log("=========================================================");

const serverProcess = spawn(nodePath, [serverJs], {
  cwd: rootDir,
  env: process.env,
  stdio: 'inherit'
});

// Tự động mở trình duyệt sau 1.5 giây
setTimeout(() => {
  console.log(`\n> Đang tự động mở trình duyệt tại http://localhost:${PORT}...`);
  exec(`start http://localhost:${PORT}`);
}, 1500);

// Đăng ký dọn dẹp tiến trình khi đóng
const cleanExit = () => {
  console.log("\nĐang dừng máy chủ Next.js và thoát...");
  serverProcess.kill('SIGINT');
  process.exit(0);
};

process.on('SIGINT', cleanExit);
process.on('SIGTERM', cleanExit);
process.on('exit', () => {
  serverProcess.kill();
});

function readlineHelper() {
  const rl = require('readline').createInterface({
    input: process.stdin,
    output: process.stdout
  });
  rl.question('Nhấn Enter để thoát...', () => {
    rl.close();
    process.exit(1);
  });
}

File: launcher.cs
using System;
using System.Diagnostics;
using System.IO;
using System.Threading;
using System.Windows.Forms;

class Program {
    static void Main() {
        string appDir = AppDomain.CurrentDomain.BaseDirectory;
        string nodeExe = Path.Combine(appDir, "bin", "node.exe");
        string serverJs = Path.Combine(appDir, "server.js");

        if (!File.Exists(serverJs)) {
            MessageBox.Show("Loi: Khong tim thay server.js trong thu muc local!\nVui long dam bao ban da giai nen dung cau truc thu muc.", "TechProject Error", MessageBoxButtons.OK, MessageBoxIcon.Error);
            return;
        }

        string nodePath = File.Exists(nodeExe) ? nodeExe : "node";

        // 1. Khoi chay Next.js server chay ngam (an hoan toan cua so console den)
        ProcessStartInfo nodeStartInfo = new ProcessStartInfo();
        nodeStartInfo.FileName = nodePath;
        nodeStartInfo.Arguments = "\"" + serverJs + "\"";
        nodeStartInfo.WorkingDirectory = appDir;
        nodeStartInfo.CreateNoWindow = true;
        nodeStartInfo.UseShellExecute = false;
        nodeStartInfo.WindowStyle = ProcessWindowStyle.Hidden;
        nodeStartInfo.EnvironmentVariables["PORT"] = "3000";
        nodeStartInfo.EnvironmentVariables["NODE_ENV"] = "production";
        nodeStartInfo.EnvironmentVariables["HOSTNAME"] = "127.0.0.1";

        Process nodeProcess = new Process();
        nodeProcess.StartInfo = nodeStartInfo;
        
        try {
            nodeProcess.Start();
        } catch (Exception ex) {
            MessageBox.Show("Khong the khoi chay may chu TechProject: " + ex.Message, "TechProject Error", MessageBoxButtons.OK, MessageBoxIcon.Error);
            return;
        }

        // Cho server Next.js khoi dong trong 1.8 giay
        Thread.Sleep(1800);

        // 2. Khoi chay Microsoft Edge o che do App Mode ( Chromium Window doc lap, khong co thanh URL )
        string edgePath = @"C:\Program Files (x86)\Microsoft\Edge\Application\msedge.exe";
        if (!File.Exists(edgePath)) {
            edgePath = "msedge.exe"; // Fallback neu nguoi dung cai o duong dan khac
        }

        string userProfileDir = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), "TechProjectProfile");
        
        ProcessStartInfo edgeStartInfo = new ProcessStartInfo();
        edgeStartInfo.FileName = edgePath;
        edgeStartInfo.Arguments = "--app=http://localhost:3000 --user-data-dir=\"" + userProfileDir + "\"";
        
        Process edgeProcess = new Process();
        edgeProcess.StartInfo = edgeStartInfo;

        try {
            edgeProcess.Start();
        } catch (Exception) {
            // Neu khong mo duoc Edge, thu mo bang trinh duyet mac dinh
            try {
                Process.Start("http://localhost:3000");
            } catch {
                MessageBox.Show("Khong the tu dong mo giao dien. Vui long truy cap http://localhost:3000 bang trinh duyet.", "TechProject Info");
            }
            nodeProcess.WaitForExit();
            return;
        }

        // 3. Cho cho den khi cua so ung dung Edge App Mode bi dong
        edgeProcess.WaitForExit();

        // 4. Khi nguoi dung dong cua so ung dung -> tu dong kill sach se tien trinh server node.exe chay ngam
        try {
            if (!nodeProcess.HasExited) {
                nodeProcess.Kill();
            }
        } catch {}
    }
}
